File: src/main/java/io/github/jolkert/discordbotlink/jda/listener/CommandHandler.java
package io.github.jolkert.discordbotlink.jda.listener;

import io.github.jolkert.discordbotlink.DiscordBotLink_LEGACY;
import io.github.jolkert.discordbotlink.jda.command.WhitelistCommand;
import io.github.jolkert.discordbotlink.util.NameUtils;
import io.github.jolkert.discordbotlink.jda.command.DiscordCommand;
import io.github.jolkert.discordbotlink.jda.command.PingCommand;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;

public class CommandHandler extends ListenerAdapter
{
	private final String prefix;
	private final HashMap<String, DiscordCommand> commandMap;
	public static final DiscordCommand[] COMMANDS = {new WhitelistCommand(), new PingCommand()};
	
	public CommandHandler(String prefix)
	{
		this.prefix	= prefix;
		
		commandMap = new HashMap<String, DiscordCommand>();
		for (DiscordCommand command : COMMANDS)
			for (String alias : command.getAllAliases())
				commandMap.put(alias, command);
	}
	
	@Override
	public void onMessageReceived(MessageReceivedEvent event)
	{
		if (event.getAuthor().isBot() || event.getAuthor().isSystem())
			return;

		if (containsPrefix(event.getMessage().getContentRaw(), prefix))
		{
			String[] args = event.getMessage().getContentRaw().substring(prefix.length()).split(" ");
			args[0] = args[0].toLowerCase();
			
			DiscordBotLink_LEGACY.Logger.info(NameUtils.nameWithDiscriminator(event.getAuthor()) + " attempting to run [" + args[0] + "]");
			if (commandMap.containsKey(args[0]))
			{
				DiscordCommand command = commandMap.get(args[0]);
				long start = System.currentTimeMillis();
				command.execute(args, event);
				DiscordBotLink_LEGACY.Logger.info("Command [" + command.getMainAlias() + "] took " + (System.currentTimeMillis() - start) + "ms");
			}
			else
				DiscordBotLink_LEGACY.Logger.info("Could not find command [" + args[0] + "]");
		}
	}
	@Override
	public void onSlashCommandInteraction(@NotNull SlashCommandInteractionEvent event)
	{
		if (!commandMap.containsKey(event.getName()))
			return;

		commandMap.get(event.getName()).executeSlashCommand(event);
	}


	private static boolean containsPrefix(String content, String prefix)
	{// TODO: eventually make this check for a mention prefix too
		return content.startsWith(prefix);
	}
}

File: src/main/kotlin/io/github/jolkert/discordbotlink/jda/listener/CommandHandler.kt
package io.github.jolkert.discordbotlink.jda.listener

import io.github.jolkert.discordbotlink.DiscordBotLink.LOGGER
import io.github.jolkert.discordbotlink.extension.nameWithDiscriminator
import io.github.jolkert.discordbotlink.jda.command.Command
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import kotlin.reflect.full.createInstance

class CommandHandler(private val commandPrefix: String) : ListenerAdapter()
{
	// TODO: try to make this a singleton object? I think that would make sense?
	companion object
	{
		val commands: Array<Command> =
			Command::class.sealedSubclasses.let { Array<Command>(it.size) { i: Int -> it[i].createInstance() } }

		private val commandMap: HashMap<String, Command> =
			HashMap<String, Command>().apply { commands.forEach { putValueAtKeys(it, it.allIdentifiers) } }

		// extensions
		private fun <K, V> HashMap<K, V>.putValueAtKeys(value: V, keys: Sequence<K>)
		{
			for (key in keys)
				this[key] = value
		}

		private fun String.afterPrefix(prefix: String): String = substring(prefix.length)
	}

	override fun onMessageReceived(event: MessageReceivedEvent)
	{
		if (event.author.isBot || event.author.isSystem || !event.message.contentRaw.startsWith(commandPrefix))
			return

		val args: Array<String> = event.message.contentRaw.afterPrefix(commandPrefix).split(' ').toTypedArray()
			.also { it[0] = it[0].lowercase() }
		LOGGER.info("${event.author.nameWithDiscriminator} attempting to run command [${args[0]}]")

		val command = commandMap[args[0]]
		if (command == null)
		{
			LOGGER.info("Could not find command [${args[0]}]")
			return
		}

		val startTime = System.currentTimeMillis()
		command.executeTextCommand(event, *args)
		LOGGER.info("Command [${command.name}] executed in ${System.currentTimeMillis() - startTime}ms")
	}

	override fun onSlashCommandInteraction(event: SlashCommandInteractionEvent)
	{
		LOGGER.info("${event.user.nameWithDiscriminator} running slash command [${event.name}]")
		val startTime = System.currentTimeMillis()
		commandMap[event.name]?.executeSlashCommand(event)
		LOGGER.info("Command [${event.name}] executed in ${System.currentTimeMillis() - startTime}ms")
	}
}